File: PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-appservice-java/src/main/java/com/microsoft/azure/toolkit/intellij/legacy/webapp/runner/webappconfig/ui/WebAppDeployConfigurationPanel.java
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 */

package com.microsoft.azure.toolkit.intellij.legacy.webapp.runner.webappconfig.ui;

import com.intellij.icons.AllIcons;
import com.intellij.ide.IdeTooltip;
import com.intellij.ide.IdeTooltipManager;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.ui.popup.Balloon;
import com.intellij.ui.HideableDecorator;
import com.intellij.ui.HyperlinkLabel;
import com.microsoft.azure.toolkit.intellij.common.AzureArtifact;
import com.microsoft.azure.toolkit.intellij.common.AzureArtifactComboBox;
import com.microsoft.azure.toolkit.intellij.common.AzureArtifactManager;
import com.microsoft.azure.toolkit.intellij.common.AzureFormPanel;
import com.microsoft.azure.toolkit.intellij.common.component.UIUtils;
import com.microsoft.azure.toolkit.intellij.legacy.appservice.AppServiceComboBox;
import com.microsoft.azure.toolkit.intellij.legacy.appservice.table.AppSettingsTable;
import com.microsoft.azure.toolkit.intellij.legacy.appservice.table.AppSettingsTableUtils;
import com.microsoft.azure.toolkit.intellij.legacy.common.AzureSettingPanel;
import com.microsoft.azure.toolkit.intellij.legacy.webapp.runner.webappconfig.IntelliJWebAppSettingModel;
import com.microsoft.azure.toolkit.intellij.legacy.webapp.runner.webappconfig.WebAppConfiguration;
import com.microsoft.azure.toolkit.lib.Azure;
import com.microsoft.azure.toolkit.lib.appservice.config.AppServiceConfig;
import com.microsoft.azure.toolkit.lib.appservice.config.DeploymentSlotConfig;
import com.microsoft.azure.toolkit.lib.appservice.config.RuntimeConfig;
import com.microsoft.azure.toolkit.lib.appservice.webapp.AzureWebApp;
import com.microsoft.azure.toolkit.lib.appservice.webapp.WebApp;
import com.microsoft.azure.toolkit.lib.appservice.webapp.WebAppDeploymentSlot;
import com.microsoft.azure.toolkit.lib.auth.AzureAccount;
import com.microsoft.azure.toolkit.lib.common.form.AzureFormInput;
import com.microsoft.azure.toolkit.lib.common.task.AzureTask;
import com.microsoft.azure.toolkit.lib.common.task.AzureTaskManager;
import org.apache.commons.collections4.MapUtils;
import org.apache.commons.compress.utils.FileNameUtils;
import org.apache.commons.lang3.ArrayUtils;
import org.apache.commons.lang3.ObjectUtils;
import org.apache.commons.lang3.StringUtils;
import org.jetbrains.idea.maven.model.MavenConstants;
import reactor.core.publisher.Mono;
import reactor.core.scheduler.Schedulers;

import javax.annotation.Nonnull;
import javax.swing.*;
import java.awt.*;
import java.awt.event.FocusEvent;
import java.awt.event.FocusListener;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.List;
import java.util.*;

import static com.microsoft.azure.toolkit.intellij.common.AzureBundle.message;

public class WebAppDeployConfigurationPanel extends AzureSettingPanel<WebAppConfiguration> implements AzureFormPanel<IntelliJWebAppSettingModel> {
    private static final String[] FILE_NAME_EXT = {"war", "jar", "ear"};
    private static final String DEPLOYMENT_SLOT = "&Deployment Slot";
    private static final String DEFAULT_SLOT_NAME = "slot-%s";
    public static final String DO_NOT_CLONE_SLOT_CONFIGURATION = "Don't clone configuration from an existing slot";
    public static final String NEW_CONFIGURATION_SOURCE = "new";

    private JPanel pnlSlotCheckBox;
    private JTextField txtNewSlotName;
    private JComboBox<Object> cbxSlotConfigurationSource;
    private JCheckBox chkDeployToSlot;
    private JCheckBox chkToRoot;
    private JPanel pnlRoot;
    private JPanel pnlSlotDetails;
    private JRadioButton rbtNewSlot;
    private JRadioButton rbtExistingSlot;
    private JComboBox<Object> cbxSlotName;
    private JPanel pnlSlot;
    private JPanel pnlSlotHolder;
    private JPanel pnlCheckBox;
    private JPanel pnlSlotRadio;
    private JLabel lblSlotName;
    private JLabel lblSlotConfiguration;
    private JCheckBox chkOpenBrowser;
    private HyperlinkLabel lblNewSlot;
    private JPanel pnlExistingSlot;
    private JButton btnSlotHover;
    private AzureArtifactComboBox comboBoxArtifact;
    private JLabel lblArtifact;
    private JLabel lblWebApp;
    private WebAppComboBox comboBoxWebApp;
    private JPanel pnlAppSettings;
    private JLabel lblAppSettings;
    private AppSettingsTable appSettingsTable;

    private final HideableDecorator slotDecorator;
    private final Project project;
    private String appSettingsKey;
    private AppServiceConfig appSettingsSource;

    public WebAppDeployConfigurationPanel(@Nonnull Project project, @Nonnull WebAppConfiguration webAppConfiguration) {
        super(project);
        this.project = project;
        this.appSettingsKey = webAppConfiguration.getAppSettingsKey();

        $$$setupUI$$$();
        comboBoxWebApp.addValueChangedListener((AzureValueChangeBiListener<AppServiceConfig>) this::onWebAppChanged);
        comboBoxArtifact.addItemListener(e -> chkToRoot.setVisible(isAbleToDeployToRoot(comboBoxArtifact.getValue())));

        final ButtonGroup slotButtonGroup = new ButtonGroup();
        slotButtonGroup.add(rbtNewSlot);
        slotButtonGroup.add(rbtExistingSlot);
        rbtExistingSlot.addItemListener(e -> toggleSlotType(true));
        rbtNewSlot.addItemListener(e -> toggleSlotType(false));
        chkDeployToSlot.addItemListener(e -> toggleSlotPanel(chkDeployToSlot.isSelected()));

        final Icon informationIcon = AllIcons.General.ContextHelp;
        btnSlotHover.setIcon(informationIcon);
        btnSlotHover.setHorizontalAlignment(SwingConstants.CENTER);
        btnSlotHover.setPreferredSize(new Dimension(informationIcon.getIconWidth(), informationIcon.getIconHeight()));
        btnSlotHover.setToolTipText(message("webapp.deploy.hint.deploymentSlot"));
        // workaround to support the hover notification
        // todo: validate whether tooltip could be read by screen reader, if so, the workaround could be removed
        btnSlotHover.addFocusListener(new FocusListener() {
            private final IdeTooltipManager instance = IdeTooltipManager.getInstance();

            @Override
            public void focusGained(FocusEvent focusEvent) {
                btnSlotHover.setBorderPainted(true);
                final JLabel tooltipContent = new JLabel(message("webapp.deploy.hint.deploymentSlot"));
                // Create an instance of IdeTooltip
                final IdeTooltip tooltip = new IdeTooltip(btnSlotHover, btnSlotHover.getLocation(), tooltipContent);
                // Set the preferred position of the tooltip
                tooltip.setPreferredPosition(Balloon.Position.above);
                // Use IdeTooltipManager to show the tooltip
                IdeTooltipManager.getInstance().show(tooltip, true);
            }

            @Override
            public void focusLost(FocusEvent focusEvent) {
                btnSlotHover.setBorderPainted(false);
                instance.dispose();
            }
        });

        final JLabel labelForNewSlotName = new JLabel("Slot Name");
        labelForNewSlotName.setLabelFor(txtNewSlotName);
        final JLabel labelForExistingSlotName = new JLabel("Slot Name");
        labelForExistingSlotName.setLabelFor(cbxSlotName);

        lblArtifact.setLabelFor(comboBoxArtifact);
        lblWebApp.setLabelFor(comboBoxWebApp);

        final DateFormat df = new SimpleDateFormat("yyMMddHHmmss");
        txtNewSlotName.setText(String.format(DEFAULT_SLOT_NAME, df.format(new Date())));

        slotDecorator = new HideableDecorator(pnlSlotHolder, DEPLOYMENT_SLOT, true);
        slotDecorator.setContentComponent(pnlSlot);
    }

    @Override
    protected void resetFromConfig(@Nonnull final WebAppConfiguration configuration) {
        final IntelliJWebAppSettingModel model = configuration.getModel();
        this.appSettingsKey = configuration.getAppSettingsKey();
        setValue(model);
    }

    @Override
    protected void apply(@Nonnull final WebAppConfiguration configuration) {
        final IntelliJWebAppSettingModel model = this.getValue();
        Optional.ofNullable(model).ifPresent(configuration::setWebAppSettingModel);
    }

    @Nonnull
    @Override
    public String getPanelName() {
        return "Deploy to Azure";
    }

    @Nonnull
    @Override
    public JPanel getMainPanel() {
        return pnlRoot;
    }

    private void onWebAppChanged(final AppServiceConfig value, final AppServiceConfig before) {
        if (value == null) {
            return;
        }
        this.loadDeploymentSlot(value);
        this.loadAppSettings(value);
    }

    private synchronized void loadAppSettings(@Nonnull AppServiceConfig value) {
        final boolean sameApp = AppServiceComboBox.isSameApp(value, appSettingsSource);
        if (sameApp) {
            return;
        }
        this.appSettingsSource = value;
        appSettingsTable.loadAppSettings(() -> isDraftWebApp(value) ? value.appSettings() :
                Optional.ofNullable(getWebApp(value)).map(WebApp::getAppSettings).orElse(Collections.emptyMap()));
    }

    // merge local app settings with remote if draft web app has been created
    private Map<String, String> loadDraftAppSettings(AppServiceConfig value) {
        final WebApp webApp = Azure.az(AzureWebApp.class).webApps(value.subscriptionId()).get(value.appName(), value.resourceGroup());
        return webApp != null && webApp.exists() ? MapUtils.putAll(ObjectUtils.firstNonNull(webApp.getAppSettings(), new HashMap<>()), value.appSettings().entrySet().toArray()) : value.appSettings();
    }

    private void setComboBoxDefaultValue(JComboBox<?> comboBox, Object value) {
        //noinspection unchecked
        UIUtils.listComboBoxItems(comboBox).stream().filter(item -> item.equals(value)).findFirst().ifPresent(defaultItem -> comboBox.setSelectedItem(value));
    }

    private AppServiceConfig getSelectedWebApp() {
        return comboBoxWebApp.getValue();
    }

    private boolean isAbleToDeployToRoot(final AzureArtifact azureArtifact) {
        final AppServiceConfig selectedWebApp = getSelectedWebApp();
        if (selectedWebApp == null || azureArtifact == null) {
            return false;
        }
        final String containerName = Optional.ofNullable(selectedWebApp.getRuntime())
                                             .map(RuntimeConfig::getWebContainer).orElse(StringUtils.EMPTY);
        final String packaging = azureArtifact.getPackaging();
        final boolean isDeployingWar = StringUtils.equalsAnyIgnoreCase(packaging, MavenConstants.TYPE_WAR, "ear");
        return isDeployingWar && StringUtils.containsAnyIgnoreCase(containerName, "tomcat", "jboss");
    }

    private void toggleSlotPanel(boolean slot) {
        final boolean isDeployToSlot = slot && (getSelectedWebApp() != null);
        rbtNewSlot.setEnabled(isDeployToSlot);
        rbtExistingSlot.setEnabled(isDeployToSlot);
        lblSlotName.setEnabled(isDeployToSlot);
        lblSlotConfiguration.setEnabled(isDeployToSlot);
        cbxSlotName.setEnabled(isDeployToSlot);
        txtNewSlotName.setEnabled(isDeployToSlot);
        cbxSlotConfigurationSource.setEnabled(isDeployToSlot);
    }

    private void toggleSlotType(final boolean isExistingSlot) {
        pnlExistingSlot.setVisible(isExistingSlot);
        pnlExistingSlot.setEnabled(isExistingSlot);
        txtNewSlotName.setVisible(!isExistingSlot);
        txtNewSlotName.setEnabled(!isExistingSlot);
        lblSlotConfiguration.setVisible(!isExistingSlot);
        cbxSlotConfigurationSource.setVisible(!isExistingSlot);
    }

    private void createUIComponents() {
        // TODO: place custom component creation code here
        lblNewSlot = new HyperlinkLabel(message("webapp.deploy.noDeploymentSlot"));
        lblNewSlot.addHyperlinkListener(e -> rbtNewSlot.doClick());

        comboBoxWebApp = new WebAppComboBox(project);
        comboBoxWebApp.reloadItems();

        comboBoxArtifact = new AzureArtifactComboBox(this.project);
        comboBoxArtifact.setFileFilter(virtualFile -> {
            final String ext = FileNameUtils.getExtension(virtualFile.getPath());
            return ArrayUtils.contains(FILE_NAME_EXT, ext);
        });
        comboBoxArtifact.reloadItems();

        appSettingsTable = new AppSettingsTable();
        pnlAppSettings = AppSettingsTableUtils.createAppSettingPanel(appSettingsTable);
    }

    private void loadDeploymentSlot(@Nonnull AppServiceConfig selectedWebApp) {
        if (isDraftWebApp(selectedWebApp)) {
            chkDeployToSlot.setEnabled(false);
            chkDeployToSlot.setSelected(false);
        } else {
            chkDeployToSlot.setEnabled(true);
            Mono.fromCallable(() -> getWebApp(selectedWebApp))
                    .map(webapp -> webapp.slots().list())
                    .subscribeOn(Schedulers.boundedElastic())
                    .subscribe(slots -> AzureTaskManager.getInstance().runLater(() -> fillDeploymentSlots(slots, selectedWebApp), AzureTask.Modality.ANY));
        }
    }

    private synchronized void fillDeploymentSlots(List<WebAppDeploymentSlot> slotList, @Nonnull final AppServiceConfig selectedWebApp) {
        final String defaultSlot = (String) cbxSlotName.getSelectedItem();
        final String defaultConfigurationSource = (String) cbxSlotConfigurationSource.getSelectedItem();
        cbxSlotName.removeAllItems();
        cbxSlotConfigurationSource.removeAllItems();
        cbxSlotConfigurationSource.addItem(DO_NOT_CLONE_SLOT_CONFIGURATION);
        cbxSlotConfigurationSource.addItem(selectedWebApp.appName());
        slotList.stream().filter(Objects::nonNull).forEach(slot -> {
            cbxSlotName.addItem(slot.getName());
            cbxSlotConfigurationSource.addItem(slot.getName());
        });
        setComboBoxDefaultValue(cbxSlotName, defaultSlot);
        setComboBoxDefaultValue(cbxSlotConfigurationSource, defaultConfigurationSource);
        final boolean existDeploymentSlot = slotList.size() > 0;
        lblNewSlot.setVisible(!existDeploymentSlot);
        cbxSlotName.setVisible(existDeploymentSlot);
    }

    @Override
    public void setValue(IntelliJWebAppSettingModel data) {
        // artifact
        Optional.ofNullable(AzureArtifactManager.getInstance(this.project).getAzureArtifactById(data.getAzureArtifactType(), data.getArtifactIdentifier()))
                .ifPresent(artifact -> comboBoxArtifact.setArtifact(artifact));
        // web app
        Optional.of(data.getConfig()).filter(app -> StringUtils.isNotBlank(app.getAppName())).ifPresent(webApp -> {
            if (Azure.az(AzureAccount.class).account().getSubscriptions().stream().noneMatch(s -> s.getId().equals(webApp.subscriptionId()))) {
                comboBoxWebApp.setValue((AppServiceConfig) null);
                return;
            }
            comboBoxWebApp.setConfigModel(webApp);
            comboBoxWebApp.setValue(c -> AppServiceComboBox.isSameApp(c, webApp));
            chkDeployToSlot.setSelected(Objects.nonNull(webApp.getSlotConfig()));
            Optional.ofNullable(webApp.getSlotConfig()).ifPresent(c -> {
                final WebApp app = getWebApp(webApp);
                final WebAppDeploymentSlot slot = Optional.ofNullable(app)
                        .map(a -> a.slots().get(c.getName(), a.getResourceGroupName())).orElse(null);
                final boolean exists = Optional.ofNullable(slot).map(WebAppDeploymentSlot::exists).orElse(false);
                if (exists) {
                    rbtExistingSlot.setSelected(true);
                    cbxSlotName.setSelectedItem(c.getName());
                } else {
                    rbtNewSlot.setSelected(true);
                    txtNewSlotName.setText(c.getName());
                    final String source = c.getConfigurationSource();
                    cbxSlotConfigurationSource.setSelectedItem(StringUtils.equals(source, NEW_CONFIGURATION_SOURCE) ? DO_NOT_CLONE_SLOT_CONFIGURATION : source);
                }
            });
            WebAppDeployConfigurationPanel.this.appSettingsSource = webApp;
            appSettingsTable.setAppSettings(webApp.appSettings());
        });
        // configuration
        chkToRoot.setSelected(data.isDeployToRoot());
        chkOpenBrowser.setSelected(data.isOpenBrowserAfterDeployment());
        slotDecorator.setOn(data.isSlotPanelVisible());
    }

    @Override
    public IntelliJWebAppSettingModel getValue() {
        final IntelliJWebAppSettingModel model = new IntelliJWebAppSettingModel();
        Optional.ofNullable(comboBoxArtifact.getValue()).ifPresent(a -> {
            model.setAzureArtifactType(a.getType());
            model.setArtifactIdentifier(a.getIdentifier());
            model.setPackaging(a.getPackaging());
        });
        final DeploymentSlotConfig slotConfig = chkDeployToSlot.isSelected() ? new DeploymentSlotConfig() : null;
        if (Objects.nonNull(slotConfig)) {
            final boolean useExistingSlot = rbtExistingSlot.isSelected();
            slotConfig.setName(useExistingSlot ? Objects.toString(cbxSlotName.getSelectedItem(), null) : txtNewSlotName.getText());
            final String source = Optional.ofNullable(cbxSlotConfigurationSource.getSelectedItem()).map(String.class::cast)
                    .map(value -> StringUtils.equalsIgnoreCase(value, DO_NOT_CLONE_SLOT_CONFIGURATION) ? NEW_CONFIGURATION_SOURCE : value).orElse(null);
            slotConfig.setConfigurationSource(useExistingSlot ? null : source);
        }
        Optional.ofNullable(comboBoxWebApp.getValue())
                .map(c -> c.toBuilder().slotConfig(slotConfig).appSettings(appSettingsTable.getAppSettings()).build())
                .ifPresent(model::setConfig);
        model.setOpenBrowserAfterDeployment(chkOpenBrowser.isSelected());
        model.setSlotPanelVisible(slotDecorator.isExpanded());
        return model;
    }

    @Override
    public List<AzureFormInput<?>> getInputs() {
        return Arrays.asList(comboBoxWebApp, comboBoxArtifact);
    }

    private WebApp getWebApp(@Nonnull final AppServiceConfig config) {
        return Azure.az(AzureWebApp.class).webApps(config.subscriptionId()).get(config.appName(), config.resourceGroup());
    }

    private boolean isDraftWebApp(@Nonnull final AppServiceConfig config) {
        return !Azure.az(AzureWebApp.class).webApps(config.subscriptionId()).exists(config.appName(), config.resourceGroup());
    }
}

File: PluginsAndFeatures/azure-toolkit-for-rider/azure-intellij-plugin-appservice-dotnet/src/main/kotlin/com/microsoft/azure/toolkit/intellij/legacy/webapp/runner/webappconfig/WebAppDeployConfigurationPanel.kt
/*
 * Copyright 2018-2023 JetBrains s.r.o. and contributors. Use of this source code is governed by the MIT license.
 */

package com.microsoft.azure.toolkit.intellij.legacy.webapp.runner.webappconfig

import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.ComboBox
import com.intellij.openapi.ui.LabeledComponent
import com.intellij.ui.components.ActionLink
import com.intellij.ui.components.JBCheckBox
import com.intellij.ui.components.JBRadioButton
import com.intellij.ui.components.JBTextField
import com.intellij.ui.dsl.builder.*
import com.intellij.ui.layout.and
import com.intellij.ui.layout.selectedValueMatches
import com.jetbrains.rider.model.publishableProjectsModel
import com.jetbrains.rider.projectView.solution
import com.jetbrains.rider.run.configurations.publishing.PublishRuntimeSettingsCoreHelper
import com.microsoft.azure.toolkit.ide.appservice.model.AzureArtifactConfig
import com.microsoft.azure.toolkit.ide.appservice.model.DeploymentSlotConfig
import com.microsoft.azure.toolkit.ide.appservice.webapp.model.WebAppConfig
import com.microsoft.azure.toolkit.ide.appservice.webapp.model.WebAppDeployRunConfigurationModel
import com.microsoft.azure.toolkit.intellij.common.AzureDotnetProjectComboBox
import com.microsoft.azure.toolkit.intellij.common.AzureFormPanel
import com.microsoft.azure.toolkit.intellij.common.component.UIUtils
import com.microsoft.azure.toolkit.intellij.common.configurationAndPlatformComboBox
import com.microsoft.azure.toolkit.intellij.common.dotnetProjectComboBox
import com.microsoft.azure.toolkit.intellij.legacy.appservice.table.AppSettingsTable
import com.microsoft.azure.toolkit.intellij.legacy.appservice.table.AppSettingsTableUtils
import com.microsoft.azure.toolkit.intellij.legacy.webapp.WebAppCreationDialog.Companion.RIDER_PROJECT_CONFIGURATION
import com.microsoft.azure.toolkit.intellij.legacy.webapp.WebAppCreationDialog.Companion.RIDER_PROJECT_PLATFORM
import com.microsoft.azure.toolkit.lib.Azure
import com.microsoft.azure.toolkit.lib.appservice.webapp.AzureWebApp
import com.microsoft.azure.toolkit.lib.appservice.webapp.WebAppDeploymentSlot
import com.microsoft.azure.toolkit.lib.common.form.AzureFormInput.AzureValueChangeBiListener
import com.microsoft.azure.toolkit.lib.common.task.AzureTask
import com.microsoft.azure.toolkit.lib.common.task.AzureTaskManager
import org.apache.commons.collections4.MapUtils
import reactor.core.publisher.Mono
import reactor.core.scheduler.Schedulers
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import javax.swing.JComboBox
import javax.swing.JPanel

class WebAppDeployConfigurationPanel(private val project: Project) : AzureFormPanel<WebAppDeployRunConfigurationModel> {
    companion object {
        private const val DEFAULT_SLOT_NAME = "slot"
    }

    private val formatter = DateTimeFormatter.ofPattern("yyMMddHHmmss")

    val panel: JPanel

    private lateinit var webAppComboBox: Cell<WebAppComboBox>
    private lateinit var dotnetProjectComboBox: Cell<AzureDotnetProjectComboBox>
    private lateinit var configurationAndPlatformComboBox: Cell<LabeledComponent<ComboBox<PublishRuntimeSettingsCoreHelper.ConfigurationAndPlatform?>>>
    private lateinit var appSettingsTable: AppSettingsTable
    private lateinit var deploymentSlotGroup: CollapsibleRow
    private lateinit var deployToSlotCheckBox: Cell<JBCheckBox>
    private lateinit var newSlotRadioButton: Cell<JBRadioButton>
    private lateinit var existingSlotRadioButton: Cell<JBRadioButton>
    private lateinit var newSlotTextField: Cell<JBTextField>
    private lateinit var existingSlotNameComboBox: Cell<ComboBox<String>>
    private lateinit var slotConfigurationSourceComboBox: Cell<ComboBox<String>>
    private lateinit var noAvailableSlotLink: Cell<ActionLink>
    private lateinit var openBrowserCheckBox: Cell<JBCheckBox>

    private val lock = Any()

    init {
        panel = panel {
            row("Web App:") {
                webAppComboBox = webAppComboBox(project)
                    .align(Align.FILL)
                    .resizableColumn()
            }
            row("Project:") {
                dotnetProjectComboBox = dotnetProjectComboBox(project) { it.canBePublishedToAzure() }
                    .align(Align.FILL)
            }
            row("Configuration:") {
                configurationAndPlatformComboBox = configurationAndPlatformComboBox(project)
                    .align(Align.FILL)
            }
            row("App Settings:") {
                appSettingsTable = AppSettingsTable()
                cell(AppSettingsTableUtils.createAppSettingPanel(appSettingsTable))
                    .align(Align.FILL)
            }
            deploymentSlotGroup = collapsibleGroup("Deployment Slot") {
                row {
                    deployToSlotCheckBox = checkBox("Deploy to slot")
                }
                buttonsGroup {
                    row {
                        newSlotRadioButton = radioButton("Create new slot")
                            .enabledIf(deployToSlotCheckBox.selected.and(webAppComboBox.component.selectedValueMatches { it != null }))
                        existingSlotRadioButton = radioButton("Use existing slot")
                            .enabledIf(deployToSlotCheckBox.selected.and(webAppComboBox.component.selectedValueMatches { it != null }))
                    }
                }
                row("Slot Name:") {
                    newSlotTextField = textField()
                        .enabledIf(deployToSlotCheckBox.selected.and(webAppComboBox.component.selectedValueMatches { it != null }))
                        .enabledIf(newSlotRadioButton.selected)
                        .visibleIf(newSlotRadioButton.selected)
                        .align(Align.FILL)

                    existingSlotNameComboBox = comboBox(emptyList<String>())
                        .enabledIf(deployToSlotCheckBox.selected.and(webAppComboBox.component.selectedValueMatches { it != null }))
                        .enabledIf(existingSlotRadioButton.selected)
                        .visibleIf(existingSlotRadioButton.selected)
                        .align(Align.FILL)
                }
                row {
                    noAvailableSlotLink = link("No available deployment slot, click to create a new one") {
                        newSlotRadioButton.component.doClick()
                    }
                        .visibleIf(existingSlotRadioButton.selected.and(existingSlotNameComboBox.component.selectedValueMatches { it == null }))
                }
                row("Slot Configuration Source:") {
                    slotConfigurationSourceComboBox = comboBox(emptyList<String>())
                        .enabledIf(deployToSlotCheckBox.selected.and(webAppComboBox.component.selectedValueMatches { it != null }))
                        .enabledIf(newSlotRadioButton.selected)
                        .align(Align.FILL)
                }
                    .enabledIf(newSlotRadioButton.selected)
                    .visibleIf(newSlotRadioButton.selected)
            }
            row {
                openBrowserCheckBox = checkBox("Open browser after deployment")
            }
        }

        webAppComboBox.component.apply {
            addValueChangedListener(AzureValueChangeBiListener(::onWebAppChanged))
            reloadItems()
        }
        dotnetProjectComboBox.component.reloadItems()
        newSlotRadioButton.component.isSelected = true
        newSlotTextField.component.text = "${DEFAULT_SLOT_NAME}-${LocalDateTime.now().format(formatter)}"
    }

    override fun setVisible(visible: Boolean) {
    }

    private fun onWebAppChanged(value: WebAppConfig, before: WebAppConfig?) {
        loadDeploymentSlot(value)
        loadAppSettings(value, before)
    }

    private fun loadDeploymentSlot(selectedWebApp: WebAppConfig) {
        val resourceId = selectedWebApp.resourceId
        if (resourceId.isNullOrEmpty()) {
            deployToSlotCheckBox.component.isEnabled = false
            deployToSlotCheckBox.component.isSelected = false
        } else {
            deployToSlotCheckBox.component.isEnabled = true
            Mono.fromCallable { Azure.az(AzureWebApp::class.java).webApp(resourceId) }
                .map { it?.slots()?.list() ?: emptyList() }
                .subscribeOn(Schedulers.boundedElastic())
                .subscribe {
                    AzureTaskManager.getInstance()
                        .runLater({ fillDeploymentSlots(it, selectedWebApp) }, AzureTask.Modality.ANY)
                }
        }
    }

    private fun fillDeploymentSlots(slotList: List<WebAppDeploymentSlot>, selectedWebApp: WebAppConfig) {
        synchronized(lock) {
            val defaultSlot = existingSlotNameComboBox.component.selectedItem as? String?
            val defaultConfigurationSource = slotConfigurationSourceComboBox.component.selectedItem as? String?
            existingSlotNameComboBox.component.removeAllItems()
            slotConfigurationSourceComboBox.component.removeAllItems()
            slotConfigurationSourceComboBox.component.addItem("Don't clone configuration from an existing slot")
            slotConfigurationSourceComboBox.component.addItem(selectedWebApp.name)
            slotList.stream().filter { it != null }.forEach {
                existingSlotNameComboBox.component.addItem(it.name)
                slotConfigurationSourceComboBox.component.addItem(it.name)
            }
            setComboBoxDefaultValue(existingSlotNameComboBox.component, defaultSlot)
            setComboBoxDefaultValue(slotConfigurationSourceComboBox.component, defaultConfigurationSource)
        }
    }

    private fun loadAppSettings(selectedWebApp: WebAppConfig, before: WebAppConfig?) {
        val rawValue =
            if (webAppComboBox.component.rawValue is WebAppConfig) webAppComboBox.component.rawValue as WebAppConfig
            else selectedWebApp
        if (before == null && selectedWebApp != rawValue) {
            if (rawValue.resourceId.isNullOrEmpty() && !selectedWebApp.resourceId.isNullOrEmpty()) {
                appSettingsTable.loadAppSettings { loadDraftAppSettings(rawValue) }
            }
        } else if (selectedWebApp != before) {
            appSettingsTable.loadAppSettings {
                if (selectedWebApp.resourceId.isNullOrEmpty()) {
                    selectedWebApp.appSettings
                } else {
                    Azure.az(AzureWebApp::class.java).webApp(selectedWebApp.resourceId)?.appSettings
                }
            }
        }
    }

    private fun loadDraftAppSettings(value: WebAppConfig): Map<String, String> {
        val webApp =
            Azure.az(AzureWebApp::class.java).webApps(value.subscriptionId).get(value.name, value.resourceGroupName)
        return if (webApp != null && webApp.exists())
            MapUtils.putAll(webApp.appSettings, value.appSettings.entries.toTypedArray())
        else
            value.appSettings
    }

    override fun setValue(data: WebAppDeployRunConfigurationModel) {
        data.artifactConfig?.let {
            val artifactId = it.artifactIdentifier?.toIntOrNull()
            if (artifactId != null) {
                project.solution.publishableProjectsModel.publishableProjects.values
                    .firstOrNull { p -> p.projectModelId == artifactId }
                    ?.let { p -> dotnetProjectComboBox.component.setProject(p) }
            }
        }

        data.webAppConfig?.let { webApp ->
            webAppComboBox.component.setConfigModel(webApp)
            webAppComboBox.component.value = webApp
            appSettingsTable.setAppSettings(webApp.appSettings)

            webApp.deploymentSlot?.let { slot ->
                deployToSlotCheckBox.component.isSelected = true
                existingSlotRadioButton.component.isSelected = !slot.isNewCreate
                if (slot.isNewCreate) {
                    newSlotTextField.component.text = slot.name
                    slotConfigurationSourceComboBox.component.addItem(slot.configurationSource)
                    slotConfigurationSourceComboBox.component.selectedItem = slot.configurationSource
                } else {
                    existingSlotNameComboBox.component.addItem(slot.name)
                    existingSlotNameComboBox.component.selectedItem = slot.name
                }
            }
        }

        openBrowserCheckBox.component.isSelected = data.isOpenBrowserAfterDeployment
        deploymentSlotGroup.expanded = data.isSlotPanelVisible
    }

    fun setConfigurationAndPlatform(configuration: String, platform: String) {
        for (i in 0 until configurationAndPlatformComboBox.component.component.model.size) {
            val item = configurationAndPlatformComboBox.component.component.model.getElementAt(i)
            if (item?.configuration == configuration && item.platform == platform) {
                configurationAndPlatformComboBox.component.component.selectedItem = item
                break
            }
        }
    }

    override fun getValue(): WebAppDeployRunConfigurationModel {
        val project = dotnetProjectComboBox.component.value
        val artifactConfig = project?.let {
            AzureArtifactConfig.builder().artifactIdentifier(it.projectModelId.toString()).build()
        }

        val slotConfig =
            if (deployToSlotCheckBox.component.isSelected) {
                if (existingSlotRadioButton.component.isSelected) {
                    DeploymentSlotConfig
                        .builder()
                        .newCreate(false)
                        .name(existingSlotNameComboBox.component.selectedItem as? String?)
                        .build()
                } else {
                    DeploymentSlotConfig
                        .builder()
                        .newCreate(true)
                        .name(newSlotTextField.component.text)
                        .configurationSource(slotConfigurationSourceComboBox.component.selectedItem as? String?)
                        .build()
                }
            } else null

        if (webAppComboBox.component.value?.appSettings?.containsKey(RIDER_PROJECT_CONFIGURATION) == true)
            webAppComboBox.component.value?.appSettings?.remove(RIDER_PROJECT_CONFIGURATION)
        if (webAppComboBox.component.value?.appSettings?.containsKey(RIDER_PROJECT_PLATFORM) == true)
            webAppComboBox.component.value?.appSettings?.remove(RIDER_PROJECT_PLATFORM)

        val webAppConfig = webAppComboBox.component.value
            ?.toBuilder()
            ?.appSettings(appSettingsTable.appSettings)
            ?.deploymentSlot(slotConfig)
            ?.build()

        return WebAppDeployRunConfigurationModel
            .builder()
            .webAppConfig(webAppConfig)
            .artifactConfig(artifactConfig)
            .openBrowserAfterDeployment(openBrowserCheckBox.component.isSelected)
            .slotPanelVisible(deploymentSlotGroup.expanded)
            .build()
    }

    fun getSelectedConfiguration() = getSelectedConfigurationAndPlatform()?.configuration ?: ""
    fun getSelectedPlatform() = getSelectedConfigurationAndPlatform()?.platform ?: ""
    private fun getSelectedConfigurationAndPlatform(): PublishRuntimeSettingsCoreHelper.ConfigurationAndPlatform? =
        configurationAndPlatformComboBox.component.component.selectedItem as? PublishRuntimeSettingsCoreHelper.ConfigurationAndPlatform

    override fun getInputs() = listOf(webAppComboBox.component, dotnetProjectComboBox.component)

    private fun setComboBoxDefaultValue(comboBox: JComboBox<*>, value: Any?) {
        UIUtils.listComboBoxItems(comboBox).stream().filter { it == value }.findFirst()
            .ifPresent { comboBox.selectedItem = value }
    }
}